src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,575 @@
+}
+
+void RemoteCacheController::deallocFile(FILE * file_stream)
+{
+    {
+        std::lock_guard lock{mutex};
+        auto it = opened_file_streams.find(file_stream);
+        if (it == opened_file_streams.end())
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Try to deallocate file with invalid handler remote path: {}, local path: {}",
+                file_meta_data.remote_path,
+                local_path.string());
+        opened_file_streams.erase(it);
+    }
+    fclose(file_stream);
+}
+
+LocalCachedFileReader::LocalCachedFileReader(RemoteCacheController * cache_controller_)
+    : cache_controller(cache_controller_)
+    , file_size(cache_controller_->getFileMetaData().file_size)
+    , offset(0)
+{
+    std::tie(file_stream, local_path) = cache_controller->allocFile();
+}
+
+LocalCachedFileReader::~LocalCachedFileReader()
+{
+    cache_controller->deallocFile(file_stream);
+}
+
+size_t LocalCachedFileReader::read(char * buf, size_t size)
+{
+    auto wret = cache_controller->waitMoreData(offset, offset + size);
+    if (wret != RemoteReadBufferCacheError::OK)
+        return 0;
+    std::lock_guard lock(mutex);
+    auto ret_size = fread(buf, 1, size, file_stream);
+    offset += ret_size;
+    return ret_size;
+}
+
+off_t LocalCachedFileReader::seek(off_t off)
+{
+    cache_controller->waitMoreData(off, 0);
+    std::lock_guard lock(mutex);
+    auto ret = fseek(file_stream, off, SEEK_SET);
+    if (ret < 0)
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS, "Seek file {} with size {} to offset {} failed: {}", getPath().string(), getSize(), off, errnoToString(errno));
+
+    offset = off;
+    return off;
+}
+
+size_t LocalCachedFileReader::getSize()
+{
+    return file_size;
+}
+
+// the size need be equal to the original buffer
+RemoteReadBuffer::RemoteReadBuffer(size_t buff_size) : BufferWithOwnMemory<SeekableReadBufferWithSize>(buff_size)
+{
+}
+
+RemoteReadBuffer::~RemoteReadBuffer() = default;
+
+std::unique_ptr<RemoteReadBuffer> RemoteReadBuffer::create(ContextPtr context, const RemoteFileMetadata & remote_file_meta, std::unique_ptr<ReadBuffer> read_buffer)
+{
+    auto * log = &Poco::Logger::get("RemoteReadBuffer");
+    size_t buff_size = DBMS_DEFAULT_BUFFER_SIZE;
+    if (read_buffer)
+        buff_size = read_buffer->internalBuffer().size();
+    /*
+     * in the new implement of ReadBufferFromHDFS, buffer size is 0.
+     *
+     * in the common case, we don't read bytes from readbuffer directly, so set buff_size = DBMS_DEFAULT_BUFFER_SIZE
+     * is OK.
+     *
+     * we need be careful with the case  without local file reader.
+     */
+    if (buff_size == 0)
+        buff_size = DBMS_DEFAULT_BUFFER_SIZE;
+
+    const auto & remote_path = remote_file_meta.remote_path;
+    auto remote_read_buffer = std::make_unique<RemoteReadBuffer>(buff_size);
+    auto * raw_rbp = read_buffer.release();
+    std::shared_ptr<ReadBuffer> srb(raw_rbp);
```

Comment:
Please change this, `srb` and `rbp` names are bad.

Reply:
done